File: ios-bookmind/Sources/SharedModels/Memory.swift
import Foundation

public struct Chunk: Identifiable, Codable, Sendable, Hashable {
    public let id: ChunkID
    public let bookID: BookID
    public let chapterID: ChapterID
    public let index: Int
    public let text: String
    public let tokenCount: Int
    public let spoilerLevel: Int

    public init(
        id: ChunkID,
        bookID: BookID,
        chapterID: ChapterID,
        index: Int,
        text: String,
        tokenCount: Int,
        spoilerLevel: Int = 0
    ) {
        self.id = id
        self.bookID = bookID
        self.chapterID = chapterID
        self.index = index
        self.text = text
        self.tokenCount = tokenCount
        self.spoilerLevel = spoilerLevel
    }
}

public struct Character: Identifiable, Codable, Sendable, Hashable {
    public let id: String
    public let bookID: BookID
    public let canonicalName: String
    public let aliases: [String]
    public let firstChapterIndex: Int
    public let lastSafeChapterIndex: Int
    public let safeSummary: String?

    public init(
        id: String,
        bookID: BookID,
        canonicalName: String,
        aliases: [String] = [],
        firstChapterIndex: Int,
        lastSafeChapterIndex: Int,
        safeSummary: String? = nil
    ) {
        self.id = id
        self.bookID = bookID
        self.canonicalName = canonicalName
        self.aliases = aliases
        self.firstChapterIndex = firstChapterIndex
        self.lastSafeChapterIndex = lastSafeChapterIndex
        self.safeSummary = safeSummary
    }
}

public struct Event: Identifiable, Codable, Sendable, Hashable {
    public let id: String
    public let bookID: BookID
    public let chapterID: ChapterID
    public let chapterIndex: Int
    public let eventType: String
    public let shortDescription: String
    public let longDescriptionSafe: String?
    public let spoilerLevel: Int

    public init(
        id: String,
        bookID: BookID,
        chapterID: ChapterID,
        chapterIndex: Int,
        eventType: String,
        shortDescription: String,
        longDescriptionSafe: String? = nil,
        spoilerLevel: Int = 0
    ) {
        self.id = id
        self.bookID = bookID
        self.chapterID = chapterID
        self.chapterIndex = chapterIndex
        self.eventType = eventType
        self.shortDescription = shortDescription
        self.longDescriptionSafe = longDescriptionSafe
        self.spoilerLevel = spoilerLevel
    }
}

public struct Relation: Identifiable, Codable, Sendable, Hashable {
    public let id: String
    public let bookID: BookID
    public let sourceCharacterID: String
    public let targetCharacterID: String
    public let relationType: String
    public let confidence: Double
    public let safeUntilChapterIndex: Int

    public init(
        id: String,
        bookID: BookID,
        sourceCharacterID: String,
        targetCharacterID: String,
        relationType: String,
        confidence: Double,
        safeUntilChapterIndex: Int
    ) {
        self.id = id
        self.bookID = bookID
        self.sourceCharacterID = sourceCharacterID
        self.targetCharacterID = targetCharacterID
        self.relationType = relationType
        self.confidence = confidence
        self.safeUntilChapterIndex = safeUntilChapterIndex
    }
}

public struct Recap: Identifiable, Codable, Sendable, Hashable {
    public let id: String
    public let bookID: BookID
    public let chapterID: ChapterID
    public let chapterIndex: Int
    public let recapText: String
    public let style: String

    public init(
        id: String,
        bookID: BookID,
        chapterID: ChapterID,
        chapterIndex: Int,
        recapText: String,
        style: String = "safe"
    ) {
        self.id = id
        self.bookID = bookID
        self.chapterID = chapterID
        self.chapterIndex = chapterIndex
        self.recapText = recapText
        self.style = style
    }
}

public struct Fact: Identifiable, Codable, Sendable, Hashable {
    public let id: String
    public let bookID: BookID
    public let chapterID: ChapterID
    public let chapterIndex: Int
    public let factType: String
    public let subjectID: String?
    public let objectID: String?
    public let text: String
    public let spoilerLevel: Int

    public init(
        id: String,
        bookID: BookID,
        chapterID: ChapterID,
        chapterIndex: Int,
        factType: String,
        subjectID: String? = nil,
        objectID: String? = nil,
        text: String,
        spoilerLevel: Int = 0
    ) {
        self.id = id
        self.bookID = bookID
        self.chapterID = chapterID
        self.chapterIndex = chapterIndex
        self.factType = factType
        self.subjectID = subjectID
        self.objectID = objectID
        self.text = text
        self.spoilerLevel = spoilerLevel
    }
}

File: android-bookmind/app/src/main/java/com/bookmind/core/model/Memory.kt
package com.bookmind.core.model

/** = iOS `Chunk`. */
data class Chunk(
    val id: ChunkID,
    val bookID: BookID,
    val chapterID: ChapterID,
    val index: Int,
    val text: String,
    val tokenCount: Int,
    val spoilerLevel: Int = 0
)

/** = iOS `Character`. */
data class Character(
    val id: String,
    val bookID: BookID,
    val canonicalName: String,
    val aliases: List<String> = emptyList(),
    val firstChapterIndex: Int,
    val lastSafeChapterIndex: Int,
    val safeSummary: String? = null
)

/** = iOS `Event`. */
data class Event(
    val id: String,
    val bookID: BookID,
    val chapterID: ChapterID,
    val chapterIndex: Int,
    val eventType: String,
    val shortDescription: String,
    val longDescriptionSafe: String? = null,
    val spoilerLevel: Int = 0
)

/** = iOS `Recap`. */
data class Recap(
    val id: String,
    val bookID: BookID,
    val chapterID: ChapterID,
    val chapterIndex: Int,
    val recapText: String,
    val style: String = "safe"
)

/** = iOS `Fact`. */
data class Fact(
    val id: String,
    val bookID: BookID,
    val chapterID: ChapterID,
    val chapterIndex: Int,
    val factType: String,
    val subjectID: String? = null,
    val objectID: String? = null,
    val text: String,
    val spoilerLevel: Int = 0
)

/** = iOS `Relation`. */
data class Relation(
    val id: String,
    val bookID: BookID,
    val sourceCharacterID: String,
    val targetCharacterID: String,
    val relationType: String,
    val confidence: Double,
    val safeUntilChapterIndex: Int
)
